src/main/java/frc/robot/commands/TrackTarget.kt

```diff
@@ -1,72 +1,210 @@
+        var h = func(nextSeed)/funcPrime(nextSeed)
+
+        while(abs(h) >= AimingConstants.EPSILON){
+
+            h = func(nextSeed)/funcPrime(nextSeed)
+            nextSeed -= h
+
+        }
+        return nextSeed
     }
 
-    fun reset(){
-        samples = Array(TrackClosest.SAMPLE_NUM){ null }
+    private fun constructTrajectory(){
+        //samples
+
+        //offset timestamps so 0 is the current time
+        val smaples = setSampleEpoch(TargetVision.smaples, TargetVision.curTime)
+
+
+        //Generalized Linear Model regresses points to find quadratic functions for translation x and y in terms of t
+        val samplerX: Sampler.DataSampleBuilder = Sampler()
+            .forColumn("t").generate{_, index -> smaples[index].timestamp}
+            .forColumn("t^2").generate{_, index -> smaples[index].timestamp.pow(2)}
+            .forColumn("x/y").generate {_, index -> smaples[index].pose.x}
+            .end()
+
+        val samplerY: Sampler.DataSampleBuilder = Sampler()
+            .forColumn("t").generate{_, index -> smaples[index].timestamp}
+            .forColumn("t^2").generate{_, index -> smaples[index].timestamp.pow(2)}
+            .forColumn("x/y").generate {_, index -> smaples[index].pose.y}
+
+            .end()
+
+        var frameX: DataFrame = schema.build()
+        var frameY: DataFrame = schema.build()
+
+        frameX = samplerX.sample(frameX, AimingConstants.SAMPLE_NUM)
+        frameY = samplerY.sample(frameY, AimingConstants.SAMPLE_NUM)
+
+
+        glm.solverType = GlmSolverType.GlmIrls
+
+        glm.fit(frameX)
+        CoefficientsList.X.coefficients.values = glm.coefficients.values
+
+        glm.fit(frameY)
+        CoefficientsList.Y.coefficients.values = glm.coefficients.values
+
+        val x = quadraticBuilder(CoefficientsList.X.coefficients)
+        val y = quadraticBuilder(CoefficientsList.Y.coefficients)
+        val vx = quadraticDerivativeBuilder(CoefficientsList.X.coefficients)
+        val vy = quadraticDerivativeBuilder(CoefficientsList.Y.coefficients)
+        val tofDerivative = quadraticDerivativeBuilder(CoefficientsList.Tof.coefficients)
+
+        val lhsfunction: (t: Double) -> Double = { t -> pointsToDistance.compose(timeOfFlight)(x(t), y(t))}
+        val lhsDerivative: (t: Double) -> Double = { t -> pointsToDistance.compose(tofDerivative)(x(t),y(t))*pointsToDistanceDerivative(vx, vy, x, y)(t)}
+
+        optimalTimeOffset = newtonRaphson(AimingConstants.INITIAL_SEED, lhsfunction, lhsDerivative)
+
+        predictedTranslation = Translation2d(vx(optimalTimeOffset), vy(optimalTimeOffset))
+    }
+
+
+
+    override fun execute() {
+        if(TargetVision.hasTargets){
+            constructTrajectory()
+            Turret.setTarget(rotationalOffset)
+        }
     }
 
+
+    private var predictedTranslation: Translation2d =  Translation2d()
+    private val curVelocity: Translation2d
+        get() = Drivetrain.velocity2d
+
+    private val rotationalOffset: Rotation2d
```

Comment:
Consider making aimandshoot aware of the turret rotation limits.

Reply:
laterrr